enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,216 @@
+- How to implement successful 3rd party software installation and upgrades
+- How to verify 3rd party software operation and provide a fallback if failed
+
+## Proposal
+
+### Workflow Description
+
+**System Administrator** is a human user responsible for managing Edge Devices.
+When a system administrator triggers a software or operating system upgrade
+on Edge Devices based on `rpm-ostree`, the changes do not take place until the
+next system reboot.
+
+**End User** is a human user operating an Edge Device. End users continue their
+work with the device running the previous software version until the reboot.
+
+Once the system is rebooted implicitly (i.e. `cron` schedule) or explicitly
+by a system administrator, 
+the [greenboot-healthcheck.service](https://github.com/fedora-iot/greenboot#how-does-it-work)
+runs to verify the system validity.
+
+If the system validity procedures are successful, the version of the software
+is declared valid for end users. Otherwise, `greenboot` enters a system
+verification loop by repeating reboots up-to `GREENBOOT_MAX_BOOT_ATTEMPTS` times
+before attempting to rollback to the previous version of the software.
+
+[![Mermaid Upgrade Flow](https://mermaid.ink/img/pako:eNpV0cFuwjAMANBfsXIdHHbtYVJRVzRpTBMMLk0PpnEhIk2iNGVDbf99oammkVMUP1uO3bPKCGIJOzm0Z_jKuIZw0mJnav-NjuD6UcJy-QLD38veBixogFWxpaMxHj5JC6lPZUxezf7WemogkgGy_oBKCvQE_2o_PY8xKYtJkxngtXgw5YN509eo8uJATtayQi-Nhndj7CzzKPcWvIEN_sxdtEP2EE-VghylIgGpD83aIGBdbI1SR6wuc7H1hFOu2YI15BqUIgysvwc582dqiLMkXAW6C2dcj8Fh583upiuWeNfRgnX2_vNMYhhdw5IaVRteSUhv3CZuYFrE-AvrnYIR?type=png)](https://mermaid.live/edit#pako:eNpV0cFuwjAMANBfsXIdHHbtYVJRVzRpTBMMLk0PpnEhIk2iNGVDbf99oammkVMUP1uO3bPKCGIJOzm0Z_jKuIZw0mJnav-NjuD6UcJy-QLD38veBixogFWxpaMxHj5JC6lPZUxezf7WemogkgGy_oBKCvQE_2o_PY8xKYtJkxngtXgw5YN509eo8uJATtayQi-Nhndj7CzzKPcWvIEN_sxdtEP2EE-VghylIgGpD83aIGBdbI1SR6wuc7H1hFOu2YI15BqUIgysvwc582dqiLMkXAW6C2dcj8Fh583upiuWeNfRgnX2_vNMYhhdw5IaVRteSUhv3CZuYFrE-AvrnYIR)
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+The MicroShift core RPM package will have a dependency on the `greenboot` RPM
+and it will include a health check script to validate that all the required
+MicroShift services are up and running. The script will be installed into the
+`/etc/greenboot/check/required.d` directory and it will not proceed with the
+validation in case the MicroShift service was not enabled or the system was not
+booted via `ostree`.
+
+> If other scripts are present in the `/etc/greenboot/check/required.d` directory,
+> they will be executed by `greenboot` in the alphanumeric order.
+
+If the `greenboot-default-health-check` RPM subpackage is installed, it already
+includes [health check scripts](https://github.com/fedora-iot/greenboot#health-checks-included-with-subpackage-greenboot-default-health-checks)
+verifying that DNS and `ostree` services can be accessed.
+
+The health check script will implement the following validations for the core
+MicroShift namespaces:
```

Comment:
It's already clarified in the table shown below, but here it seems healthchecks are only implemented on additional infra services. Maybe state that healthchecks are implemented on core MicroShift state and its services (running as pods on ns)?

Reply:
Done